### PR TITLE
Reorganisation and clean-up performed.

### PR DESCRIPTION
**Pull request by rashtrakoff**
1. There were some functions that were common to both the strategies and will probably be common to all the future strategies as well. These functions have been moved to an abstract contract called `StrategyBase`.

2. Support for supertoken check statement has been removed from the `topUp` functions and instead put in the `checkTopUpByIndex` function of `StrollManager`. `checkTopUpByIndex` is a view function which will be called by external keepers to check if a topup is necessary. This will save us gas costs as earlier, this check was being performed in the `topUp` function which isn't a view function.

3. `require` statements have been converted to `error` statements to reduce gas costs and more informative revert statements. Some `error` statements now contain parameters that provide more information regarding revert reasons.

4. Tests have been modified to work with the custom errors. However, instead of checking the custom error that's being thrown,
we are now checking if the function is reverting or not i.e., without any specific reason.

5. NatSpec comments have been added wherever necessary (in the interfaces). However, the same remains to be done in the `IStrollManager` interface. @harsh132 please add the same from your side with necessary comments in the `StrollManager` contract as well.

Further changes to `StrollManager` must be avoided. If any feature needs to be added, note down the same and we will add it in the next version of the contracts.

### CODE REVIEW

File: smart-contracts/contracts/StrollManager.sol [hunk 1]
```diff
@@ -1,56 +1,21 @@
 //SPDX-License-Identifier: Unlicense
-pragma solidity ^0.8.4;
+pragma solidity =0.8.13;
```
Comment (ngmachado) [inline smart-contracts/contracts/StrollManager.sol:2]: is enough to declare 0.8.13

Comment (rashtrakoff) [reply smart-contracts/contracts/StrollManager.sol:2]: Isn't this a bit more precise though? Have seen this used somewhere.

**Review by ngmachado — APPROVED** [review]
LGTM

**Comment by rashtrakoff** [comment]
There are a couple of changes that needs to be done:

- Remove personal email. That will invite spam emails. Suggest using completely different email like Proton Mail.
- Move the NatSpec comments from main contract (`StrollManager`) to the interface `IStrollManager`. Use pointer (`IStrollManager.createTopUp implementation`) in the main contract.

Apart from this, I don't see any issues.

**Pull request merged**